tensorflow-framework/src/test/java/org/tensorflow/framework/metrics/AUCTest.java

```diff
@@ -0,0 +1,326 @@
+/* Copyright 2020 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+=======================================================================*/
+package org.tensorflow.framework.metrics;
+
+import org.junit.jupiter.api.Test;
+import org.tensorflow.Operand;
+import org.tensorflow.framework.utils.TestSession;
+import org.tensorflow.op.Op;
+import org.tensorflow.op.Ops;
+import org.tensorflow.types.TFloat32;
+import org.tensorflow.types.TInt32;
+import org.tensorflow.types.TInt64;
+
+import static org.junit.jupiter.api.Assertions.assertArrayEquals;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.tensorflow.framework.utils.CastHelper.cast;
+
+public class AUCTest {
+  private final TestSession.Mode tfMode = TestSession.Mode.GRAPH;
+
+  float epsilon = 1e-4F;
+
+  int numThresholds = 3;
+  float[] predArray = new float[] {0f, 0.5f, 0.3f, 0.9f};
+  int[] trueArray = new int[] {0, 0, 1, 1};
+  float[] sampleWeight = new float[] {1, 2, 3, 4};
+
+  @Test
+  public void testValueIsIdempotent() {
+    try (TestSession session = TestSession.createTestSession(tfMode)) {
+      Ops tf = session.getTF();
+      Operand<TFloat32> yPred = tf.constant(predArray);
+      Operand<TInt32> yTrue = tf.constant(trueArray);
+      AUC<TFloat32> instance = new AUC<>(tf, numThresholds, 1001L, TFloat32.class);
+
+      session.run(tf.init());
+
+      Op update = instance.updateState(yTrue, yPred, null);
+
+      for (int i = 0; i < 10; i++) {
+        session.run(update);
+      }
+
+      Operand<TFloat32> result = instance.result();
+
+      for (int i = 0; i < 10; i++) {
+        session.evaluate(result, instance.result());
```

Comment:
Similarly, what effect do we expect from these repeated calls to `session.evaluate`? What would be the equivalent in production code? Are we confident that `session.evaluate` accurately mirrors that production equivalent? Is there a way for us to write our code more like production code?

Reply:
Again this is testing that the first fetch of `result`, does not change after subsequent `result`  fetches.  Again, different implementation of result logic might produce different results when `result` is evaluated due to hidden behavior (e.g. randomness) in the TF Operations.